File: library system/books/AuthorController.js
import { v4 as uuidv4 } from 'uuid';
function newuuid() {
    const id = uuidv4();
  return id;
}
export function createAuthor(name){
    const headers = new Headers()
    const options = {
        method: "POST",
        headers: headers,
        body: JSON.stringify({
            "name":name,
        })
        //mode: 'no-cors'
    };
    const uri = "http://localhost:8080/authors"
    
    
    return fetch(uri, options).then(response => response.json())
            
    
}

export function getAuthor(){
    const headers = new Headers()
    const options = {
        method: "GET",
        headers: headers,
        //mode: 'no-cors'
    };
    const uri = "http://localhost:8080/authors"
    
    return fetch(uri, options).then(response => response.json())
    
}

export function getAuthorById(id){
    const headers = new Headers()
    const options = {
        method: "GET",
        headers: headers,
        //mode: 'no-cors'
    };
    const uri = "http://localhost:8080/authors/" + id
    
    return fetch(uri, options).then(response => response.json())
    
}

export function updateAuthor(id,author){
    const headers = new Headers()
    const options = {
        method: "PUT",
        headers: headers,
        body: JSON.stringify(author)
        //mode: 'no-cors'
    };
    const uri = "http://localhost:8080/authors/" + id
    
    return fetch(uri, options).then(response => response.json())
    
}

export function deleteAuthors(id){
    const headers = new Headers()
    const options = {
        method: "DELETE",
        headers: headers,
        //mode: 'no-cors'
    };
    const uri = "http://localhost:8080/authors/" + id
    
    return fetch(uri, options).then(response => response.json())
    
}

File: library system/books/AuthorController.go
package controller

import (
	"log"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type Author struct {
	Id   uuid.UUID `gorm:"primaryKey" json:"id"`
	Name string    `json:"name"`
}

func GetAuthors(c *gin.Context) {
	db := connectDB()
	var authors *[]Author
	db.Find(&authors)
	closeDB(db)
	c.JSON(200, authors)
}

func GetAuthorById(c *gin.Context) {
	db := connectDB()
	var author *Author
	queryResult := db.Where("id = $1", c.Param("id")).Take(&author)
	if queryResult.Error != nil {
		c.JSON(500, gin.H{
			"message": "查無此人" + queryResult.Error.Error(),
		})
		closeDB(db)
		return
	}
	closeDB(db)
	c.JSON(200, author)
}

func CreateAuthor(c *gin.Context) {
	db := connectDB()
	var author *Author
	c.BindJSON(&author)
	result := db.Create(&author)
	if result.Error != nil {
		log.Println(result.Error)
		c.JSON(500, gin.H{
			"message": "新增失敗" + result.Error.Error(),
		})
		closeDB(db)
		return
	}
	closeDB(db)
	c.JSON(200, author)
}

func UpdateAuthor(c *gin.Context) {
	db := connectDB()
	var author *Author

	queryResult := db.Where("id = $1", c.Param("id")).Take(&author)
	if queryResult.Error != nil {
		log.Println(queryResult.Error)
		c.JSON(500, gin.H{
			"message": "查無此人" + queryResult.Error.Error(),
		})
		closeDB(db)
		return
	}
	var authorbody *Author
	c.BindJSON(&authorbody)
	authorbody.Id = author.Id

	result := db.Model(&author).Where("id = ?", author.Id).Updates(&authorbody)

	if result.Error != nil {
		log.Println(result.Error)
		c.JSON(500, gin.H{
			"message": "更新失敗" + result.Error.Error(),
		})
		closeDB(db)
		return
	}
	closeDB(db)
	c.JSON(200, author)
}

func DeleteAuthor(c *gin.Context) {
	db := connectDB()
	var author *Author

	queryResult := db.Where("id = $1", c.Param("id")).Take(&author)
	if queryResult.Error != nil {
		log.Println(queryResult.Error)
		c.JSON(500, gin.H{
			"message": "查無此人" + queryResult.Error.Error(),
		})
		closeDB(db)
		return
	}

	result := db.Delete(&author)

	if result.Error != nil {
		log.Println(result.Error)
		c.JSON(500, gin.H{
			"message": "刪除失敗" + result.Error.Error(),
		})
		closeDB(db)
		return
	}
	closeDB(db)
	c.JSON(200, author)
}

